loader/loader.go

```
@@ -0,0 +1,591 @@
+		return errors.Trace(err)
+	}
+
+	l.jobWg.Wait()
+
+	err = l.checkPoint.Save(dataFile)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	log.Infof("[loader][saved to checkpoint]")
+	log.Infof("[loader][restore table data sql]%s/%s[finished]", path, dataFile)
+	return nil
+}
+
+func causeErr(err error) error {
+	var e error
+	for {
+		e = errors.Cause(err)
+		if err == e {
+			break
+		} else {
+			err = e
+		}
+	}
+	return e
+}
+
+func (l *Loader) restoreData() error {
+	var err error
+	l.conns, err = createConns(l.cfg.DB, l.cfg.Worker)
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	l.wg.Add(l.cfg.Worker)
+	for i := 0; i < l.cfg.Worker; i++ {
+		go l.runWorker(l.conns[i], l.jobsQueue)
+	}
+
+	// restore db in sort
+	dbs := make([]string, 0, len(l.dbTables))
+	for db, _ := range l.dbTables {
+		dbs = append(dbs, db)
+	}
+	sort.Strings(dbs)
+	for _, db := range dbs {
+		tables := l.dbTables[db]
+		// create db
+		dbFile := fmt.Sprintf("%s/%s-schema-create.sql", l.cfg.Dir, db)
+		log.Infof("[loader][run db schema]%s[start]", dbFile)
+		err = l.restoreSchema(l.conns[0], dbFile, "")
+		if err != nil {
+			err = causeErr(err)
+			if e, ok := err.(*mysql.MySQLError); ok && e.Number == tidb.ErrDBCreateExists {
+				log.Warnf("run db schema failed - %v", errors.ErrorStack(err))
+			} else {
+				log.Fatalf("run db schema failed - %v", errors.ErrorStack(err))
+			}
+		}
+		log.Infof("[loader][run db schema]%s[finished]", dbFile)
+
+		// restore table in sort
+		tnames := make([]string, 0, len(tables))
+		for t, _ := range tables {
+			tnames = append(tnames, t)
+		}
+		sort.Strings(tnames)
+		for _, table := range tnames {
+			// create table
+			tableFile := fmt.Sprintf("%s/%s.%s-schema.sql", l.cfg.Dir, db, table)
+			log.Infof("[loader][run table schema]%s[start]", tableFile)
+			err := l.restoreSchema(l.conns[0], tableFile, db)
+			if err != nil {
+				err = causeErr(err)
+				if e, ok := err.(*mysql.MySQLError); ok && e.Number == tidb.ErrTableExists {
```

Tell you another way to do it (not suggest you change current implementation):

Go have a `context.Context` , you can pass a context to restoreSchema, set a context variable like "ignoreExistError"...

Then you can let the callee, i.e, `restoreSchema`  to decide return error or not, rather than the caller check all kinds of return errors, find the special error, do additional action